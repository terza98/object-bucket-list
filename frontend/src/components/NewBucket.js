import React, {useState, useEffect} from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Container from 'react-bootstrap/Container';
import Button from 'react-bootstrap/Button';

//API import
import Service from '../api/auth-service.js'

export default function NewBucket(props){
    const [bucketLocations, loadBucketLocation] = useState([""]);
    const [isLoaded, setLoad] = useState(false);
    const [error, setError] = useState("");
    const [name, setName] = useState("");
    const [location, setLocation] = useState("");

    useEffect(() => {
        // Update the bucket location
        Service.getBucketLocation()
            .then(
                (result) => {
                    setLoad(true);
                    loadBucketLocation(result.data.locations);
                },
                // Note: hanling errors here
                (error) => {
                    setError(error);
                }
            )
    }, [] );

    function handleNameChange(e){
        setName(e.target.value);
    }
    function handleLocationChange(e){
        setLocation(e.target.value);
        console.log(e.target.value);
    }

    return(
        props.new &&
        <>
            <p className="text-left">{props.title}</p>
            <Container className="innerWrapper">
                <Form>
                    <Row>
                        <Col className="text-left">
                            <Form.Label className="required">Bucket Name</Form.Label>
                            <Form.Control onChange={handleNameChange} value={name} placeholder="MyNewStorage" />
                        </Col>
                        <Col className="text-left">
                            <Form.Label className="required">Bucket Location</Form.Label>
                            <Form.Control onChange={handleLocationChange} value={location} as="select">
                                <option value="0">Choose...</option>
                                {bucketLocations.map( (item,index) =>
                                    <option key={index} value={item.id}>{item.name}</option>
                                )}
                            </Form.Control>
                        </Col>
                    </Row>  
                    <Row className="text-left">
                        {props.new&&
                            <Button style={{margin: '15px'}} onClick={() => props.handleNewBucket(name, location)}>Create Bucket</Button>  
                        }   
                    </Row>
                </Form>
            </Container>
        </>
    );
}