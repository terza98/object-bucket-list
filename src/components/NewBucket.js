import React, { useState, useEffect } from 'react';

// https://react-bootstrap.github.io/getting-started/introduction#importing-components
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Container from 'react-bootstrap/Container';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';

//API import
import ApiClient from '../api/auth-service.js';

export default function NewBucket(props) {
	const [bucketLocations, loadBucketLocation] = useState(['']);
	const [name, setName] = useState('');
	const [location, setLocation] = useState('');

	//general state
	const [loaded, setLoaded] = useState(false);
	const [error, setError] = useState('');

	useEffect(() => {
		// Update the bucket location
		ApiClient.getBucketLocation()
			.then(result => {
				loadBucketLocation(result.data.locations);
			})
			.catch(error => {
				setError(error.message);
			})
			.finally(() => setLoaded(true));
	}, []);

	function handleNameChange(e) {
		setName(e.target.value);
	}
	function handleLocationChange(e) {
		setLocation(e.target.value);
	}

	if (!loaded) {
		return <img src={require('../loading.gif')} alt="loading..." />;
	}
	return (
		<>
			<p className="text-left">Create New Bucket</p>
			<Container className="innerWrapper">
				{error && (
					<Alert
						variant="danger"
						onClose={() => setError('')}
						dismissible
					>
						{typeof error !== 'object' ? error : ''}
					</Alert>
				)}
				<Form
					onSubmit={e => {
						e.preventDefault();
						props.onSubmit(name, location);
					}}
				>
					<Row>
						<Col className="text-left">
							<Form.Label className="required">
								Bucket Name
							</Form.Label>
							<Form.Control
								onChange={handleNameChange}
								value={name}
								data-testid="input"
								placeholder="MyNewStorage"
							/>
						</Col>
						<Col className="text-left">
							<Form.Label className="required">
								Bucket Location
							</Form.Label>
							<Form.Control
								onChange={handleLocationChange}
								value={location}
								data-testid="select"
								as="select"
							>
								<option value="0">Choose...</option>
								{bucketLocations !== null &&
									bucketLocations.map((item, index) => (
										<option key={index} value={item.id}>
											{item.name}
										</option>
									))}
							</Form.Control>
						</Col>
					</Row>
					<Row className="text-left">
						<Button
							style={{ margin: '15px' }}
							type="submit"
							data-testid="button"
						>
							Create Bucket
						</Button>
					</Row>
				</Form>
			</Container>
		</>
	);
}
