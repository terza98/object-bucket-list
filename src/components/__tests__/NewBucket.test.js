import React from 'react';
import NewBucket from '../NewBucket.js';
import ReactDOM from 'react-dom';
import { render, fireEvent } from '@testing-library/react';
import { unmountComponentAtNode } from 'react-dom';

let container = null;
beforeEach(() => {
	// setup a DOM element as a render target
	container = document.createElement('div');
	document.body.appendChild(container);
});

afterEach(() => {
	// cleanup on exiting
	unmountComponentAtNode(container);
	container.remove();
	container = null;
});

it('renders without crashing', () => {
	ReactDOM.render(<NewBucket />, container);
});

it('matches snapshot', () => {
	const { asFragment } = render(<NewBucket />);

	expect(asFragment()).toMatchSnapshot();
});

test('should submit when clicking submit button', () => {
	const handleSubmit = jest.fn();
	const { getByTestId } = render(<NewBucket onSubmit={handleSubmit} />);

	const button = getByTestId('button');

	fireEvent.click(button);

	expect(handleSubmit).toHaveBeenCalled();
});
