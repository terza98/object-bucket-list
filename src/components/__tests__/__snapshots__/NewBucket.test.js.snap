// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <p
    class="text-left"
  >
    Create New Bucket
  </p>
  <div
    class="innerWrapper container"
  >
    <form
      class=""
    >
      <div
        class="row"
      >
        <div
          class="text-left col"
        >
          <label
            class="required form-label"
          >
            Bucket Name
          </label>
          <input
            class="form-control"
            placeholder="MyNewStorage"
            value=""
          />
        </div>
        <div
          class="text-left col"
        >
          <label
            class="required form-label"
          >
            Bucket Location
          </label>
          <select
            class="form-control"
          >
            <option
              value="0"
            >
              Choose...
            </option>
            <option />
          </select>
        </div>
      </div>
      <div
        class="text-left row"
      >
        <button
          class="btn btn-primary"
          data-testid="button"
          style="margin: 15px;"
          type="button"
        >
          Create Bucket
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
